File: typescript/src/ships/internal/loadout-fitting.ts
/** Module-to-mount compatibility rules for {@link ShipLoadout}. @internal */

import type { OutfittingModule } from '../modules.js';
import { getShipByName, getShipBySymbol } from '../ships.js';
import { SLOT_RESTRICTION_LABELS, type BuildSlot, type SlotRestriction } from '../slots.js';
import { truncate } from '../../internal/argument-guards.js';
import type { LoadoutIssueParams, ModuleFitConstraint } from '../loadout-validation.js';

/** Optional-internal groups a military slot accepts (symbol prefixes). */
const MILITARY_PREFIXES: readonly string[] = [
    'int_hullreinforcement',
    'int_metaalloyhullreinforcement',
    'int_modulereinforcement',
    'int_shieldcellbank',
    'int_guardianhullreinforcement',
    'int_guardianmodulereinforcement',
    'int_guardianshieldreinforcement',
];

/**
 * Weapon groups a mining hardpoint accepts.
 *
 * The Sub-Surface Extraction Missile is included because both source registries file
 * it with the displacement missile it is a variant of, despite its unrelated symbol.
 * The Pulse Wave Analyser is excluded because it is a utility fitting.
 */
const MINING_PREFIXES: readonly string[] = [
    'hpt_mininglaser', // Mining Laser, Mining Lance
    'hpt_mining_abrblstr', // Abrasion Blaster
    'hpt_mining_seismchrgwarhd', // Seismic Charge Launcher
    'hpt_mining_subsurfdispmisle', // Sub-Surface Displacement Missile
    'hpt_human_extraction', // Sub-Surface Extraction Missile
    'hpt_miningtoolv2', // Mining Volley Repeater
];

/**
 * Optional-internal groups a dedicated cargo slot accepts. A fuel tank counts here,
 * as it does in an unrestricted optional slot.
 */
const CARGO_PREFIXES: readonly string[] = [
    'int_cargorack',
    'int_largecargorack',
    'int_corrosionproofcargorack',
    'int_fueltank',
];

/**
 * The module symbol families accepted by each restricted mount.
 *
 * This is the mount's half of compatibility. A module reserved for one mount carries
 * its own `restrictedToSlot`, checked separately below. Labels come from the public
 * slot model so application menus and refusal messages cannot drift apart.
 */
const RESTRICTED_SLOT_PREFIXES: Record<SlotRestriction, readonly string[]> = {
    mining: MINING_PREFIXES,
    military: MILITARY_PREFIXES,
    cargo: CARGO_PREFIXES,
    // Single- and multi-limpet controllers are separate symbol families.
    limpetController: ['int_dronecontrol', 'int_multidronecontrol'],
    // One family covers both the Mk I and Mk II vessel bays.
    vesselHangar: ['int_fighterbay'],
    // Mk II cabins are a separate family, not PassengerCabin variants.
    passenger: ['int_passengercabin', 'int_mkii_passengercabin'],
    // Both ordinary and advanced suites share this prefix and reserve this mount.
    planetaryApproachSuite: ['int_planetapproachsuite'],
};

/** A fitting failure's stable code and English fallback. */
export interface ModuleFitProblem {
    readonly constraint: ModuleFitConstraint;
    readonly message: string;
    readonly params?: LoadoutIssueParams;
}

const problem = (
    constraint: ModuleFitConstraint,
    message: string,
    params?: LoadoutIssueParams,
): ModuleFitProblem => ({ constraint, message, ...(params === undefined ? {} : { params }) });

function restrictionProblem(slot: BuildSlot, symbol: string): ModuleFitProblem | null {
    const restriction = slot.restriction;
    if (!restriction) return null;
    if (RESTRICTED_SLOT_PREFIXES[restriction].some((prefix) => symbol.startsWith(prefix))) {
        return null;
    }
    return problem('restrictedMount', `slot only takes ${SLOT_RESTRICTION_LABELS[restriction]}`, {
        restriction,
    });
}

function moduleSlotProblem(slot: BuildSlot, module: OutfittingModule): ModuleFitProblem | null {
    const required = module.restrictedToSlot;
    if (!required || slot.restriction === required) return null;
    return problem(
        'restrictedMount',
        `module only fits a mount that takes ${SLOT_RESTRICTION_LABELS[required]}`,
        { restriction: required },
    );
}

/**
 * Explain why `module` cannot fit `slot`, or return `null` when it fits.
 *
 * The checks intentionally proceed from fixed hull constraints to module-specific
 * mount constraints, slot kind and finally size so callers receive the most specific
 * useful failure.
 */
export function moduleFitProblem(
    shipSymbol: string,
    slot: BuildSlot,
    module: OutfittingModule,
): ModuleFitProblem | null {
    if (slot.kind === 'cargoHatch') {
        return problem('immutableSlot', 'the cargoHatch slot cannot be changed');
    }
    if (slot.kind === 'armour') {
        const hull = getShipBySymbol(shipSymbol);
        if (module.slot !== 'armour' || module.ship === undefined) {
            return problem('armourRequired', 'not a ship armour module');
        }
        if (!hull || module.ship.toLowerCase() !== hull.name.toLowerCase()) {
            const armourHull = getShipByName(module.ship);
            return problem(
                'wrongHullArmour',
                `armour belongs to ${truncate(module.ship)}, not ${truncate(hull?.name ?? shipSymbol)}`,
                {
                    armourShipName: module.ship,
                    ...(armourHull === null ? {} : { armourShipSymbol: armourHull.symbol }),
                    shipSymbol,
                    ...(hull === null ? {} : { shipName: hull.name }),
                },
            );
        }
        return null;
    }

    // `ship` is unreliable for restricted modules (some carry a "None" sentinel), so
    // use the normalized restriction list and retain symbols in the error for journal
    // searches.
    const restricted = module.restrictedToShips;
    if (
        restricted &&
        !restricted.some((symbol) => symbol.toLowerCase() === shipSymbol.toLowerCase())
    ) {
        const allowedShipNames = restricted.map((symbol) => {
            const hull = getShipBySymbol(symbol);
            return hull?.name ?? symbol;
        });
        const labels = restricted.map((symbol, index) => {
            const name = allowedShipNames[index]!;
            return name === symbol ? symbol : `${name} (${symbol})`;
        });
        return problem('restrictedHull', `module is restricted to ${truncate(labels.join(', '))}`, {
            allowedShipNames,
            allowedShipSymbols: [...restricted],
            shipSymbol,
        });
    }

    // Check the article's reserved mount before general kind rules: this remains true
    // even when the mount offered by the caller is otherwise unrestricted.
    const wrongMount = moduleSlotProblem(slot, module);
    if (wrongMount) return wrongMount;
    const symbol = module.symbol.toLowerCase();
    // The normalized slot field handles Guardian core modules filed as `internal` and
    // hull-specific modules whose symbols share no family prefix.
    const moduleSlot = module.slot;

    switch (slot.kind) {
        case 'core':
            if (moduleSlot !== slot.core) {
                return problem('wrongCoreType', `not a ${slot.core} module`, {
                    requiredCore: slot.core,
                    moduleSlot: moduleSlot ?? 'none',
                });
            }
            break;
        case 'hardpoint': {
            if (module.category !== 'hardpoint') {
                return problem('hardpointRequired', 'not a hardpoint weapon');
            }
            const restricted = restrictionProblem(slot, symbol);
            if (restricted) return restricted;
            break;
        }
        case 'utility':
            if (module.category !== 'utility') {
                return problem('utilityRequired', 'not a utility module');
            }
            return null;
        case 'optional': {
            const isFuelTank = moduleSlot === 'fuelTank';
            if (module.category !== 'internal' && !isFuelTank) {
                return problem('optionalInternalRequired', 'not an optional-internal module');
            }
            if (moduleSlot && !isFuelTank) {
                return problem('coreModuleInOptionalSlot', 'a core module only fits its core slot');
            }
            const restricted = restrictionProblem(slot, symbol);
            if (restricted) return restricted;
            break;
        }
    }

    return module.class > slot.size
        ? problem(
              'oversized',
              `module size ${truncate(module.class)} exceeds slot size ${slot.size}`,
              { moduleClass: module.class, slotSize: slot.size },
          )
        : null;
}

/** Explain why `module` cannot fit `slot`, or return `null` when it fits. */
export function moduleFitError(
    shipSymbol: string,
    slot: BuildSlot,
    module: OutfittingModule,
): string | null {
    return moduleFitProblem(shipSymbol, slot, module)?.message ?? null;
}

File: dotnet/src/EliteDangerousAlmanac/Ships/Internal/LoadoutFitting.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Globalization;
using EliteDangerousAlmanac.Internal;

namespace EliteDangerousAlmanac.Ships.Internal;

/// <summary>The module-to-mount compatibility rules a build reads.</summary>
internal static class LoadoutFitting
{
    /// <summary>The optional-internal families a military mount accepts.</summary>
    private static readonly string[] MilitaryPrefixes =
    [
        "int_hullreinforcement",
        "int_metaalloyhullreinforcement",
        "int_modulereinforcement",
        "int_shieldcellbank",
        "int_guardianhullreinforcement",
        "int_guardianmodulereinforcement",
        "int_guardianshieldreinforcement",
    ];

    /// <summary>The weapon families a mining hardpoint accepts.</summary>
    /// <remarks>
    /// The Sub-Surface Extraction Missile is here because both source registries file it with
    /// the displacement missile it is a variant of, despite its unrelated symbol. The Pulse
    /// Wave Analyser is not, because it is a utility fitting.
    /// </remarks>
    private static readonly string[] MiningPrefixes =
    [
        "hpt_mininglaser",
        "hpt_mining_abrblstr",
        "hpt_mining_seismchrgwarhd",
        "hpt_mining_subsurfdispmisle",
        "hpt_human_extraction",
        "hpt_miningtoolv2",
    ];

    /// <summary>
    /// The optional-internal families a dedicated cargo mount accepts. A fuel tank counts
    /// here, as it does in an unrestricted optional mount.
    /// </summary>
    private static readonly string[] CargoPrefixes =
    [
        "int_cargorack",
        "int_largecargorack",
        "int_corrosionproofcargorack",
        "int_fueltank",
    ];

    /// <summary>The module families each restricted mount accepts.</summary>
    /// <remarks>
    /// This is the mount's half of compatibility. A module reserved for one mount carries its
    /// own restriction, which is checked separately.
    /// </remarks>
    private static readonly ReadOnlyDictionary<SlotRestriction, string[]> RestrictedPrefixes =
        new(new Dictionary<SlotRestriction, string[]>
        {
            [SlotRestriction.Mining] = MiningPrefixes,
            [SlotRestriction.Military] = MilitaryPrefixes,
            [SlotRestriction.Cargo] = CargoPrefixes,

            // Single-limpet and multi-limpet controllers are separate symbol families.
            [SlotRestriction.LimpetController] = ["int_dronecontrol", "int_multidronecontrol"],

            // One family covers both the Mark I and the Mark II vessel bays. Both Large
            // Planetary Vehicle Hangar lines are separate families the game also fits here — a
            // journal module purchase puts a Mark II large buggy bay into the Type-11's first
            // fighter bay, over the vessel hangar it replaced — so the mount takes three
            // families rather than one. The large buggy bay prefix covers that line's granted
            // articles too, since they fit wherever their sold twins do.
            [SlotRestriction.VesselHangar] = ["int_fighterbay", "int_largebuggybay", "int_mkiilargebuggybay"],

            // Mark II cabins are a separate family rather than passenger-cabin variants.
            [SlotRestriction.Passenger] = ["int_passengercabin", "int_mkii_passengercabin"],

            // Both the ordinary and the advanced suite share this prefix and reserve this mount.
            [SlotRestriction.PlanetaryApproachSuite] = ["int_planetapproachsuite"],
        });

    /// <summary>Why a module cannot fit a mount, or <see langword="null"/> when it fits.</summary>
    /// <remarks>
    /// The checks run from the fixed hull mount, through the articles that belong to the hull,
    /// to the module's own mount reservation, then the mount kind and finally the size, so a
    /// caller is given the most specific useful refusal.
    /// </remarks>
    internal static ModuleFitProblem? Problem(string shipSymbol, BuildSlot slot, OutfittingModule module)
    {
        if (slot.Kind == SlotKind.CargoHatch)
        {
            return new ModuleFitProblem(
                ModuleFitConstraint.ImmutableSlot, "the cargo hatch mount cannot be changed");
        }

        // The hatch arrives with the hull, no station sells it, and the one mount that holds it
        // is the fixed mount refused just above. So it fits nowhere a caller can fit it, and an
        // outfitting list must never offer it. The source registry files it under the internal
        // category, which an unrestricted optional mount would otherwise accept.
        if (LoadoutState.IsBuiltInHullSymbol(module.Symbol))
        {
            return new ModuleFitProblem(
                ModuleFitConstraint.BuiltInHullModule,
                "the cargo hatch is part of the hull, not an outfitting module");
        }

        if (slot.Kind == SlotKind.Armour) return ArmourProblem(shipSymbol, module);

        ModuleFitProblem? restrictedHull = RestrictedHullProblem(shipSymbol, module);
        if (restrictedHull is not null) return restrictedHull;

        // The article's own reserved mount comes before the general kind rules, because it
        // holds even when the mount a caller offered is otherwise unrestricted.
        if (module.RestrictedToSlot is SlotRestriction required && slot.Restriction != required)
        {
            return new ModuleFitProblem(
                ModuleFitConstraint.RestrictedMount,
                "module only fits a mount that takes " + required.Label())
            {
                Restriction = required,
            };
        }

        ModuleFitProblem? kind = KindProblem(slot, module);
        if (kind is not null) return kind;

        // A utility mount takes any utility fitting, whatever its class says.
        if (slot.Kind == SlotKind.Utility) return null;

        return module.Class > slot.Size
            ? new ModuleFitProblem(
                ModuleFitConstraint.Oversized,
                string.Format(
                    CultureInfo.InvariantCulture,
                    "module size {0} exceeds slot size {1}",
                    module.Class,
                    slot.Size))
            {
                ModuleClass = module.Class,
                SlotSize = slot.Size,
            }
            : null;
    }

    private static ModuleFitProblem? ArmourProblem(string shipSymbol, OutfittingModule module)
    {
        Ship? hull = ShipCatalogue.FindBySymbol(shipSymbol);
        if (module.Slot != ModuleSlot.Armour || module.Ship is null)
        {
            return new ModuleFitProblem(
                ModuleFitConstraint.ArmourRequired, "not a ship armour module");
        }

        if (hull is not null && RegistryIndex.KeyComparer.Equals(module.Ship, hull.Name))
        {
            return null;
        }

        Ship? armourHull = ShipCatalogue.FindByName(module.Ship);
        return new ModuleFitProblem(
            ModuleFitConstraint.WrongHullArmour,
            string.Format(
                CultureInfo.InvariantCulture,
                "armour belongs to {0}, not {1}",
                TextPreview.Truncate(module.Ship),
                TextPreview.Truncate(hull?.Name ?? shipSymbol)))
        {
            ArmourShipName = module.Ship,
            ArmourShipSymbol = armourHull?.Symbol,
            ShipSymbol = shipSymbol,
            ShipName = hull?.Name,
        };
    }

    /// <summary>
    /// Whether a module sold for named hulls only is on one of them.
    /// </summary>
    /// <remarks>
    /// The module's own ship field is unreliable for a restricted module, some carrying a
    /// sentinel, so the normalized restriction list decides. The refusal keeps the symbols as
    /// well as the names, so a reader can search a journal for either.
    /// </remarks>
    private static ModuleFitProblem? RestrictedHullProblem(string shipSymbol, OutfittingModule module)
    {
        IReadOnlyList<string>? restricted = module.RestrictedToShips;
        if (restricted is null) return null;

        foreach (string symbol in restricted)
        {
            if (RegistryIndex.KeyComparer.Equals(symbol, shipSymbol)) return null;
        }

        List<string> names = new(restricted.Count);
        List<string> symbols = new(restricted.Count);
        List<string> labels = new(restricted.Count);
        foreach (string symbol in restricted)
        {
            string name = ShipCatalogue.FindBySymbol(symbol)?.Name ?? symbol;
            names.Add(name);
            symbols.Add(symbol);
            labels.Add(name == symbol
                ? symbol
                : string.Format(CultureInfo.InvariantCulture, "{0} ({1})", name, symbol));
        }

        return new ModuleFitProblem(
            ModuleFitConstraint.RestrictedHull,
            "module is restricted to " + TextPreview.Truncate(string.Join(", ", labels)))
        {
            AllowedShipNames = new ReadOnlyCollection<string>(names),
            AllowedShipSymbols = new ReadOnlyCollection<string>(symbols),
            ShipSymbol = shipSymbol,
        };
    }

    private static ModuleFitProblem? KindProblem(BuildSlot slot, OutfittingModule module)
    {
        switch (slot.Kind)
        {
            case SlotKind.Core:
                // The normalized mount field handles the Guardian core modules the registry
                // files as internal, and the hull-specific modules whose symbols share no
                // family prefix.
                return module.Slot is ModuleSlot fitted && (int)fitted == (int)slot.Core!.Value
                    ? null
                    : new ModuleFitProblem(
                        ModuleFitConstraint.WrongCoreType,
                        string.Format(
                            CultureInfo.InvariantCulture, "not a {0} module", slot.Core))
                    {
                        RequiredCore = slot.Core,
                        ModuleSlot = module.Slot,
                    };

            case SlotKind.Hardpoint:
                return module.Category != ModuleCategory.Hardpoint
                    ? new ModuleFitProblem(
                        ModuleFitConstraint.HardpointRequired, "not a hardpoint weapon")
                    : RestrictionProblem(slot, module.Symbol);

            case SlotKind.Utility:
                return module.Category != ModuleCategory.Utility
                    ? new ModuleFitProblem(
                        ModuleFitConstraint.UtilityRequired, "not a utility module")
                    : null;

            default:
                return OptionalProblem(slot, module);
        }
    }

    private static ModuleFitProblem? OptionalProblem(BuildSlot slot, OutfittingModule module)
    {
        bool fuelTank = module.Slot == ModuleSlot.FuelTank;
        if (module.Category != ModuleCategory.Internal && !fuelTank)
        {
            return new ModuleFitProblem(
                ModuleFitConstraint.OptionalInternalRequired, "not an optional-internal module");
        }

        if (module.Slot is not null && !fuelTank)
        {
            return new ModuleFitProblem(
                ModuleFitConstraint.CoreModuleInOptionalSlot,
                "a core module only fits its core slot")
            {
                ModuleSlot = module.Slot,
            };
        }

        return RestrictionProblem(slot, module.Symbol);
    }

    private static ModuleFitProblem? RestrictionProblem(BuildSlot slot, string symbol)
    {
        if (slot.Restriction is not SlotRestriction restriction) return null;

        string wanted = symbol.ToLowerInvariant();
        foreach (string prefix in RestrictedPrefixes[restriction])
        {
            if (wanted.StartsWith(prefix, StringComparison.Ordinal)) return null;
        }

        return new ModuleFitProblem(
            ModuleFitConstraint.RestrictedMount, "slot only takes " + restriction.Label())
        {
            Restriction = restriction,
        };
    }
}
